app/src/main/java/at/roteskreuz/stopcorona/screens/reporting/reportStatus/ReportingStatusViewModel.kt

```diff
@@ -1,38 +1,107 @@
 import at.roteskreuz.stopcorona.model.repositories.QuarantineRepository
 import at.roteskreuz.stopcorona.model.repositories.ReportingRepository
 import at.roteskreuz.stopcorona.skeleton.core.model.helpers.AppDispatchers
 import at.roteskreuz.stopcorona.skeleton.core.model.helpers.DataState
 import at.roteskreuz.stopcorona.skeleton.core.model.helpers.DataStateObserver
+import at.roteskreuz.stopcorona.skeleton.core.model.helpers.State
 import at.roteskreuz.stopcorona.skeleton.core.screens.base.viewmodel.ScopedViewModel
+import at.roteskreuz.stopcorona.skeleton.core.utils.observeOnMainThread
+import com.google.android.gms.common.api.ApiException
+import com.google.android.gms.common.api.Status
+import com.google.android.gms.nearby.exposurenotification.ExposureNotificationStatusCodes
+import com.google.android.gms.nearby.exposurenotification.TemporaryExposureKey
 import io.reactivex.Observable
 import io.reactivex.rxkotlin.Observables
+import io.reactivex.rxkotlin.plusAssign
 import kotlinx.coroutines.launch
 import org.threeten.bp.ZonedDateTime
+import timber.log.Timber
 
 /**
  * Handles the user interaction and provides data for [ReportingStatusFragment].
  */
 class ReportingStatusViewModel(
     appDispatchers: AppDispatchers,
     private val reportingRepository: ReportingRepository,
-    private val quarantineRepository: QuarantineRepository
+    private val quarantineRepository: QuarantineRepository,
+    private val exposureNotificationRepository: ExposureNotificationRepository
 ) : ScopedViewModel(appDispatchers) {
-
     private val uploadReportDataStateObserver = DataStateObserver<MessageType>()
+    private val exposureNotificationsErrorState = DataStateObserver<Pair<Status, ExposureNotificationRepository.DebugAction>>()
 
     fun setUserAgreement(agreement: Boolean) {
         reportingRepository.setUserAgreement(agreement)
     }
 
     fun uploadInfectionInformation() {
+        exposureNotificationsErrorState.loading()
+        disposables += exposureNotificationRepository.observeTemporaryExposureKeys()
+            .observeOnMainThread()
+            .subscribe{
+                when (it){
+                    is DataState.Loaded ->{
+                        uploadData(it.data)
+                        exposureNotificationsErrorState.idle()
+                    }
+                    is State.Loading -> { /**nothing to do here**/ }
+                    is State.Error -> {
+
+                        val exception = it.error
+                        if (exception !is ApiException) {
+                            Timber.e(exception, "Unknown error when attempting to start API")
+                            return@subscribe
+                        }
+                        val apiException = exception
+                        if (apiException.statusCode == ExposureNotificationStatusCodes.DEVELOPER_ERROR) {
+                            disposables += exposureNotificationRepository.observeRegistrationState()
```

Comment:
I would strongly suggest you to refactor whole this method to use coroutines instead of rx observables. The flow is hardly readable and you are not disposing of old connected listeners when you are in the next phase of the flow.